File: scraper/scrapers/la_rebaja.py
"""
Scraper para La Rebaja Virtual (larebajavirtual.com).
Estrategia: VTEX catalog REST API directa.
No requiere Playwright.
"""

import httpx
from .base import BaseScraper, ScrapedProduct
from .farmatodo import _classify, _extract_concentration, _extract_presentation
from .utils import normalize

SEARCH_URL = "https://www.larebajavirtual.com/api/catalog_system/pub/products/search/"

BASE_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/124.0.0.0 Safari/537.36",
    "Accept": "application/json",
    "Origin": "https://www.larebajavirtual.com",
    "Referer": "https://www.larebajavirtual.com/",
}


def _spec(product: dict, key: str) -> str:
    """Extrae el primer valor de una especificacion VTEX."""
    values = product.get(key, [])
    return values[0].strip() if values else ""


def _map_product(p: dict) -> ScrapedProduct | None:
    name = p.get("productName", "").strip()
    if not name:
        return None

    items = p.get("items", [])
    if not items:
        return None
    sellers = items[0].get("sellers", [])
    if not sellers:
        return None
    offer = sellers[0].get("commertialOffer", {})

    list_price = offer.get("ListPrice") or 0
    sale_price = offer.get("Price") or 0
    if sale_price <= 0 or sale_price > 5_000_000:
        return None

    price = int(sale_price)
    if list_price and list_price > sale_price:
        ref_price = int(list_price)
        discount = round((1 - price / ref_price) * 100)
    else:
        ref_price = None
        discount = None

    avail_qty = offer.get("AvailableQuantity", 0) or 0
    availability = "available" if avail_qty > 0 else "unavailable"

    ingredient = _spec(p, "Principio activo") or _spec(p, "Principio Activo") or name.split()[0]
    presentation = _spec(p, "Presentacion") or _extract_presentation(name)
    concentration = _extract_concentration(name)

    qty_str = _spec(p, "Cantidadunidadesmedida")
    try:
        quantity = int(float(qty_str)) if qty_str else 1
    except ValueError:
        quantity = 1
    quantity = max(quantity, 1)

    price_per_unit = price // quantity

    is_rx = _spec(p, "Producto RX").upper() == "SI"
    product_type = _classify(not is_rx, name)

    url = p.get("link", "") or f"https://www.larebajavirtual.com/{p.get('linkText', '')}/p"

    return ScrapedProduct(
        pharmacy_id="la-rebaja",
        product_name=name,
        type=product_type,
        active_ingredient=ingredient,
        concentration=concentration,
        presentation=presentation,
        quantity=quantity,
        price=price,
        price_per_unit=price_per_unit,
        reference_price=ref_price,
        discount_pct=discount,
        availability=availability,
        url=url,
    )


class LaRebajaScraper(BaseScraper):
    PHARMACY_ID = "la-rebaja"

    def __init__(self):
        self._client = httpx.Client(follow_redirects=True, headers=BASE_HEADERS, timeout=20)

    async def search(self, query: str) -> list[ScrapedProduct]:
        results: list[ScrapedProduct] = []
        try:
            # VTEX devuelve maximo 50 por llamada; dos paginas cubren 100 productos
            for offset in range(0, 100, 50):
                r = self._client.get(
                    SEARCH_URL,
                    params={"ft": query, "_from": offset, "_to": offset + 49},
                )
                if r.status_code not in (200, 206):
                    break
                products = r.json()
                if not products:
                    break
                for p in products:
                    try:
                        product = _map_product(p)
                        if product:
                            results.append(product)
                    except Exception:
                        continue
                if len(products) < 50:
                    break
        except Exception as e:
            print(f"[la-rebaja] Error en API: {e}")

        # Filtrar: el query debe aparecer en nombre o ingrediente activo (normalizado sin acentos)
        q = normalize(query)
        results = [r for r in results if q in normalize(r.product_name) or q in normalize(r.active_ingredient)]

        print(f"[la-rebaja] '{query}' -> {len(results)} productos (VTEX API)")
        return results

File: app/lib/scrapers/la_rebaja.ts
import type { ScrapedProduct } from './types'
import { extractConcentration, extractPresentation, extractPackQuantity, LIQUID_PRESENTATIONS, classify, matchesQuery } from './utils'
import { withCache } from './cache'

const SEARCH_URL = 'https://www.larebajavirtual.com/api/catalog_system/pub/products/search/'
const BASE_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/124.0.0.0 Safari/537.36',
  'Accept': 'application/json',
  'Origin': 'https://www.larebajavirtual.com',
  'Referer': 'https://www.larebajavirtual.com/',
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function spec(p: Record<string, any>, key: string): string {
  const values = p[key]
  if (Array.isArray(values) && values.length > 0) return String(values[0]).trim()
  return ''
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function mapProduct(p: Record<string, any>): ScrapedProduct | null {
  const name = (String(p.productName ?? '')).trim()
  if (!name) return null

  const items = (p.items ?? []) as Record<string, unknown>[]
  if (!items.length) return null
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const sellers = ((items[0] as any).sellers ?? []) as Record<string, unknown>[]
  if (!sellers.length) return null
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const offer = ((sellers[0] as any).commertialOffer ?? {}) as Record<string, number>

  const salePrice = Number(offer.Price) || 0
  if (salePrice <= 0 || salePrice > 5_000_000) return null

  const price     = Math.round(salePrice)
  const listPrice = Number(offer.ListPrice) || 0
  const refPrice  = listPrice > salePrice ? Math.round(listPrice) : undefined
  const discount  = refPrice ? Math.round((1 - price / refPrice) * 100) : undefined
  const rawQty    = offer.AvailableQuantity
  const availability: ScrapedProduct['availability'] =
    rawQty === undefined ? 'available' : Number(rawQty) === 0 ? 'unavailable' : Number(rawQty) < 5 ? 'limited' : 'available'

  const ingredient = spec(p, 'Principio activo') || spec(p, 'Principio Activo') || name.split(/\s/)[0]

  // The "Presentacion" spec is frequently the packaging ("CAJA"), not the dosage
  // form. Prefer the form parsed from the name; only use the spec if it is itself
  // a real dosage form (running it through extractPresentation drops "caja" etc.).
  const presentation = extractPresentation(name) || extractPresentation(spec(p, 'Presentacion'))

  // Numeric part of the concentration (e.g. 500 from "500mg") so we can detect
  // when a spec wrongly reports the dose as the unit count.
  const concNum = parseInt(extractConcentration(name, presentation)) || 0

  // Cantidadunidadesmedida means different things per form:
  //  - volume forms (jarabe, crema, frasco): the ml/g CONTENT of one container
  //  - solids: the unit COUNT in the pack
  const specVal = parseInt(spec(p, 'Cantidadunidadesmedida')) || 0
  let quantity: number
  if (LIQUID_PRESENTATIONS.has(presentation)) {
    // Show the net content in ml/g. Cremas/lociones carry no ml in the name, so the
    // spec is the only source; prefer the name when it does state the volume.
    const fromName = extractPackQuantity(name, presentation)
    quantity = fromName > 1 ? fromName : (specVal >= 2 && specVal <= 5000 ? specVal : 1)
  } else {
    // Solids: trust the count only when sane and not just an echo of the dose number.
    if (specVal >= 2 && specVal <= 1000 && specVal !== concNum) {
      quantity = specVal
    } else {
      quantity = extractPackQuantity(name, presentation)
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const firstImage = (items[0] as any).images?.[0]?.imageUrl as string | undefined

  return {
    pharmacyId:     'la-rebaja',
    productName:    name,
    type:           classify({ name }),
    activeIngredient: ingredient,
    concentration:  extractConcentration(name, presentation),
    presentation,
    quantity:       Math.max(quantity, 1),
    price,
    pricePerUnit:   Math.round(price / Math.max(quantity, 1)),
    referencePrice: refPrice,
    discountPct:    discount,
    availability,
    url:      (p.link as string) || `https://www.larebajavirtual.com/${p.linkText ?? ''}/p`,
    imageUrl: firstImage || undefined,
  }
}

export async function searchLaRebaja(query: string): Promise<ScrapedProduct[]> {
  return withCache('la-rebaja', query, async () => {
    const results: ScrapedProduct[] = []
    let failed = false
    try {
      // Single page of 50 is plenty for relevance and avoids a second sequential
      // round-trip (each request can take several seconds).
      const params = new URLSearchParams({ ft: query, _from: '0', _to: '49' })
      const res = await fetch(`${SEARCH_URL}?${params}`, { headers: BASE_HEADERS, signal: AbortSignal.timeout(7_000) })
      if (res.status !== 200 && res.status !== 206) {
        failed = true
      } else {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const products = (await res.json()) as Record<string, any>[]
        if (Array.isArray(products)) {
          for (const p of products) {
            const product = mapProduct(p)
            if (product) results.push(product)
          }
        }
      }
    } catch (e) {
      console.error('[la-rebaja] Error:', e)
      if (results.length === 0) failed = true
    }
    if (failed) return null // failure -> serve stale cache

    return results.filter(r => matchesQuery(query, r.productName, r.activeIngredient))
  })
}
